SILOptimizer: restructure the apply(partial_apply) peephole and the dead partial_apply elimination optimizations

eeckstein opened the pull request:
Changes:

* Allow optimizing partial_apply capturing opened existential: we didn't do this originally because it was complicated to insert the required alloc/dealloc_stack instructions at the right places. Now we have the StackNesting utility, which makes this easier.

* Support indirect-in parameters. Not super important, but why not? It's also easy to do with the StackNesting utility.

* Share code between dead closure elimination and the apply(partial_apply) optimization. It's a bit of refactoring and allowed to eliminate some code which is not used anymore.

* Fix an ownership problem: We inserted copies of partial_apply arguments _after_ the partial_apply (which consumes the arguments).

* When replacing an apply(partial_apply) -> apply and the partial_apply becomes dead, avoid inserting copies of the arguments twice.

These changes don't have any immediate effect on our current benchmarks, but will allow eliminating curry thunks for existentials.

eeckstein committed:
SILOptimizer: restructure the apply(partial_apply) peephole and the dead partial_apply elimination optimizations

Changes:

* Allow optimizing partial_apply capturing opened existential: we didn't do this originally because it was complicated to insert the required alloc/dealloc_stack instructions at the right places. Now we have the StackNesting utility, which makes this easier.

* Support indirect-in parameters. Not super important, but why not? It's also easy to do with the StackNesting utility.

* Share code between dead closure elimination and the apply(partial_apply) optimization. It's a bit of refactoring and allowed to eliminate some code which is not used anymore.

* Fix an ownership problem: We inserted copies of partial_apply arguments _after_ the partial_apply (which consumes the arguments).

* When replacing an apply(partial_apply) -> apply and the partial_apply becomes dead, avoid inserting copies of the arguments twice.

These changes don't have any immediate effect on our current benchmarks, but will allow eliminating curry thunks for existentials.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci benchmark

gottesmm commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -295,16 +295,36 @@ struct InstModCallbacks {
+/// Returns true, if there are no other users beside those collected in \p
+/// destroys, i.e. if \p inst can be considered as "dead".
+bool collectDestroys(SingleValueInstruction *inst,
+                             SmallVectorImpl<SILInstruction *> &destroys);
```
Can you run this patch through git-clang-format?

eeckstein replied:
okay, done

gottesmm reviewed (commented):
Quick nit. I need to read this offline in my editor. I'll do a more in depth in a bit.

futurejones commented:
Can one of the admins verify this patch?

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci smoke test

gottesmm commented on lib/SILOptimizer/Utils/ValueLifetime.cpp:
```diff
@@ -281,3 +289,20 @@ void ValueLifetimeAnalysis::dump() const {
+
+void swift::endLifetimeAtFrontier(
+    SILValue valueOrStackLoc, const ValueLifetimeAnalysis::Frontier &frontier,
+    SILBuilder &builder) {
```
Why pass in a SILBuilder here? Better to pass in a SILBuilderContext and use a locally created SILBuilderWithScope. Also, why are you messing with RegularLocations here?

eeckstein replied:
SILBuilder: ok, makes sense.

The frontier instruction can have a ReturnLocation. Therefore I have to construct a RegularLocation.

gottesmm commented on lib/SILOptimizer/Utils/ValueLifetime.cpp:
```diff
@@ -190,7 +197,8 @@ bool ValueLifetimeAnalysis::computeFrontier(Frontier &frontier, Mode mode,
     for (unsigned i = 0, e = succBlocks.size(); i != e; ++i) {
       if (unhandledFrontierBlocks.count(succBlocks[i])) {
-        assert(isCriticalEdge(term, i) && "actually not a critical edge?");
+        assert((isCriticalEdge(term, i) || userSet.count(term)) &&
```
Can you explain this to me? Why is this change needed? Does userSet.count(term) mean a critical edge?

eeckstein replied:
This is a bug fix.
```
/// An edge is also considered as "critical" if it has a single precedessor
/// but the predecessor's terminal instruction is a user of the value.
///
```

gottesmm commented on lib/SILOptimizer/Utils/ValueLifetime.cpp:
```diff
@@ -135,8 +135,15 @@ bool ValueLifetimeAnalysis::computeFrontier(Frontier &frontier, Mode mode,
-      // frontier (see below).
-      assert(deadInSucc && "The final using TermInst must have successors");
+      // frontier (see below). Except it is a function exiting TermInst, like
+      // 'return' or 'throw'.
```
Can you expand the comment here why the term inst is an exception. I understand why, it should just be explicit in the code.

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
-  }
-
-  builder.setInsertionPoint(insertPoint);
+  builder.setInsertionPoint(paiAI.getInstruction());
```
Please change this to use a new SILBuilderWithScope that takes in a SILBuilderContext.

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -229,78 +139,80 @@ bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
+    auto *apply = cast<ApplyInst>(paiAI);
+    auto *newAI = builder.createApply(paiAI.getLoc(), callee, subs, argList,
+                                      apply->isNonThrowing());
+    callbacks.replaceValueUsesWith(apply, newAI);
```
Did you add this callback due to a test failure? Which test was it?

eeckstein replied:
This callback was already in the original code.

gottesmm replied:
I see. It was hoisted from below.

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -347,25 +259,29 @@ SILInstruction *PartialApplyCombiner::combine() {
-  if (!tmpCopies.empty()) {
-    destroyTemporaries();
-    deallocateTemporaries();
+  if (foundApplySites.empty())
```
I like the way this loop is done here. Much better!

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
-    args[i] = createLifetimeExtendedAllocStack(builder, loc, arg, exitingBlocks,
-                                               callbacks);
+    SILType ty = argOp.get()->getType();
+    if (!ty.isTrivial(*F) || (includeTrivialAddrArgs && ty.isAddress()))
```
This is really nice here!

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
+    if (useHasTransitiveOwnership(user)) {
+      if (!collectDestroys(cast<SingleValueInstruction>(user), destroys))
+        isDead = false;
+      destroys.push_back(user);
```
Rather than else if, can you use early continues and a worklist instead of recursion?

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
+
+  for (Operand *argOp : argsToHandle) {
+    SILValue arg = argOp->get();
+    int argIdx = argOp->getOperandNumber() - pai->getArgumentOperandNumber();
```
Please do not hard code this directly. This should definitely be something on function convention or apply site apply or something like that. @atrick your thoughts?

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
+    SILValue arg = argOp->get();
+    int argIdx = argOp->getOperandNumber() - pai->getArgumentOperandNumber();
+    SILDebugVariable dbgVar(/*Constant*/ true, argIdx);
+    builder.setInsertionPoint(pai);
```
We are trying to move away from code like this. Can you use a new SILBuilderWithScope if possible?

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1521,38 +1510,53 @@ bool swift::tryDeleteDeadClosure(SingleValueInstruction *closure,
-      return false;
+    assert(!pa->isOnStack() &&
+           "partial_apply [stack] should have been handled before");
+    SILBuilder builder(pai);
```
SILBuilder => SILBuilderWithScope

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1521,38 +1510,53 @@ bool swift::tryDeleteDeadClosure(SingleValueInstruction *closure,
+    if (needKeepArgsAlive) {
+      if (!keepArgsOfPartialApplyAlive(pai, closureDestroys, builder))
+        return false;
+    } else {
```
No else after return according to LLVM style guide. Please remove this.

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1521,38 +1510,53 @@ bool swift::tryDeleteDeadClosure(SingleValueInstruction *closure,
+      for (Operand *argOp : argsToHandle) {
+        SILValue arg = argOp->get();
+        if (arg->getType().isObject()) {
+          builder.emitDestroyValueOperation(pai->getLoc(), arg);
```
I think we do this a lot. Add this as a helper on SILBuilder? Maybe emitDestroyOperation?

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
+      tmp = builder.emitCopyValueOperation(pai->getLoc(), arg);
+    } else {
+      // Copy address-arguments into a stack-allocated temporary.
+      tmp = builder.createAllocStack(pai->getLoc(), arg->getType(), dbgVar);
```
I am fine with this now, but I do think that at some point, we should incorporate code from SILGen in terms of ManagedValue/Scopes here. Would make it much easier to work with.

gottesmm commented on include/swift/SILOptimizer/Utils/ValueLifetime.h:
```diff
@@ -79,6 +79,9 @@ class ValueLifetimeAnalysis {
   /// the method getCriticalEdges can be used to retrieve the critical edges.
   ///
+  /// An edge is also considered as "critical" if it has a single precedessor
+  /// but the predecessor's terminal instruction is a user the value.
```
Can you elaborate on this here in the comment. The comment is defining something without explaining why. An example would be useful.

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -229,78 +139,80 @@ bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
-        toBeDestroyedArgs.push_back(arg);
+      if (arg->getType().isAddress()) {
+        auto *ASI = builder.createAllocStack(pai->getLoc(), arg->getType());
+        builder.createCopyAddr(pai->getLoc(), arg, ASI, IsTake_t::IsNotTake,
```
Thanks for fixing this. We should be copying before each call site. This would be another place where SILGen's ManagedValue/Scope machinery would be useful.

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -229,78 +139,80 @@ bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
-    nai = builder.createApply(paiAI.getLoc(), callee, subs, argList,
-                              cast<ApplyInst>(paiAI)->isNonThrowing());
-
+  if (auto *tai = dyn_cast<TryApplyInst>(paiAI)) {
```
Can you change this to be a full switch over the type of full apply site with llvm_unreachable in the cases which we do not support today. That way people know to update this code in the future.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -342,7 +342,8 @@ static SILValue cleanupLoadedCalleeValue(SILValue calleeValue, LoadInst *li) {
 /// Removes instructions that create the callee value if they are no
 /// longer necessary after inlining.
-static void cleanupCalleeValue(SILValue calleeValue) {
+static void cleanupCalleeValue(SILValue calleeValue,
```
Why not return needUpdateStackNesting as a bool?

gottesmm commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -104,10 +104,18 @@ SILInstruction *SILCombiner::visitPartialApplyInst(PartialApplyInst *PAI) {
+  // need to copy all arguments again (to extend their lifetimes), because it
+  // was already done in tryOptimizeApplyOfPartialApply.
+  if (tryDeleteDeadClosure(PAI, getInstModCallbacks(), !argsAreKeptAlive))
+    needUpdateStackNesting = true;
```
needUpdateStackNesting => stackNestingInvalidated?

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
+  // Compute the set of endpoints, which will be used to insert destroys of
+  // temporaries. This may fail if the frontier is located on a critical edge
+  // which we may not split.
+  ValueLifetimeAnalysis vla(pai, paiUsers);
```
Thank you for fixing this. I was worried about this code using ValueLifetimeAnalysis(SILValue) API. We should really get rid of that entry point since it is so fragile and easy to mess up.

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
-  SILBuilderWithScope builder(pai);
-  SILLocation loc = pai->getLoc();
+static bool
+deadMarkDependenceUser(SILInstruction *inst,
```
Is this dead now? I noticed you got rid of the usage below?

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
+///       not be needed anymore with OSSA.
+static bool keepArgsOfPartialApplyAlive(PartialApplyInst *pai,
+                                        ArrayRef<SILInstruction *> paiUsers,
+                                        SILBuilder &builder) {
```
Another case where I would suggest passing in a SILBuilderContext and creating builders locally. Will prevent Debug Info bugs.

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
-    argsToHandle.push_back(std::make_pair(arg, i));
+  // Of course we must inlude all apply instructions which we want to optimize.
+  for (FullApplySite ai : applies) {
+    paiUsers.push_back(ai.getInstruction());
```
What about this:

```
transform(applies, std::back_inserter(paiUsers), [](FullApplySite ai) { return ai.getInstruction(); })
```

eeckstein replied:
I find the for-loop version easier to understand.

gottesmm replied:
Ok.

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
+/// as they may be destroyed/deallocated before the last use by one of the
+/// apply instructions.
+bool PartialApplyCombiner::copyArgsToTemporaries(
+    const SmallVectorImpl<FullApplySite> &applies) {
```
=> ArrayRef<FullApplySite>

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
-      continue;
+  ArrayRef<SILParameterInfo> paramList = paiTy->getParameters();
+  auto argList = pai->getArgumentOperands();
+  paramList = paramList.drop_front(paramList.size() - argList.size());
```
Can we make an API around this on ApplySite? Or on partial apply?

gottesmm commented on lib/SILOptimizer/Utils/PartialApplyCombiner.cpp:
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
-      continue;
+  for (Operand *argOp : argsToHandle) {
+    SILValue arg = argOp->get();
+    int argIdx = argOp->getOperandNumber() - pai->getArgumentOperandNumber();
```
Please don't hard code this. This should be sort sort of higher level API on function conventions or ApplySite.

eeckstein commented:
@gottesmm Thanks for the review. I have addressed your points in this new version. Much of the code is just copied from the original version, which is already quite old. That's why SILBuilderContext etc. was not used. But, yes, it makes sense to refresh the code.

eeckstein commented:
@swift-ci smoke test

gottesmm commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -104,10 +104,20 @@ SILInstruction *SILCombiner::visitPartialApplyInst(PartialApplyInst *PAI) {
     return nullptr;
 
-  tryOptimizeApplyOfPartialApply(PAI, Builder, getInstModCallbacks());
+  SILBuilderContext BuilderCtxt(Builder.getModule(), Builder.getTrackingList());
```
You can get this directly from the builder.

vedantk commented:
Hi @eeckstein @gottesmm, I'm starting to see a use-before-init on the sanitizer bot after this PR:

```
/Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/swift/lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:122:9: runtime error: load of value 109, which is not a valid value for type 'bool'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/swift/lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:122:9 in 
Stack dump:
0.	Program arguments: /Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/Ninja-ReleaseAssert+asan+ubsan/swift-macosx-x86_64/bin/swift -frontend -module-cache-path /Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/Ninja-ReleaseAssert+asan+ubsan/lldb-macosx-x86_64/./lldb-test-build.noindex/module-cache-clang -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -serialize-debugging-options -module-cache-path /Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/Ninja-ReleaseAssert+asan+ubsan/lldb-macosx-x86_64/test/Swift/Output/RemoteASTImport.test.tmp/cache -merge-modules -emit-module -parse-as-library -sil-merge-partial-modules -disable-diagnostic-passes -disable-sil-perf-optzns -module-name Library Library.part.swiftmodule -o Library.swiftmodule -I/Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/Ninja-ReleaseAssert+asan+ubsan/lldb-macosx-x86_64/test/Swift/Output/RemoteASTImport.test.tmp 
```

This is here:

```
  bool runOnFunction(SILFunction &function) {
    bool changed = false;

    while (doOneIteration(function, iteration)) {
      changed = true;
      ++iteration;
    }

    if (invalidatedStackNesting) { //< here
      StackNesting().correctStackNesting(&function);
    }
```

vedantk commented:
This also shows up when building the swift stdlib, so no lldb testing required to repro:

```
3.      While running pass #1 SILFunctionTransform "MandatoryCombine" on SILFunction "@$sSAyxGs8_PointerssABP9_rawValueBpvgTW".
 for getter for _rawValue (at /Users/vsk/src/github-swift-master/swift/stdlib/public/core/BridgeObjectiveC.swift:408:14)
```

eeckstein commented:
@vedantk Thanks for letting me know.
https://github.com/apple/swift/pull/29842 should fix the issue
